SharkCage/CageServiceInstaller/SharkCageInstallerHelper.cs

```diff
@@ -0,0 +1,143 @@
+﻿using Microsoft.Win32;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections;
+using System.ComponentModel;
+using System.Configuration.Install;
+using System.IO;
+using System.Security.AccessControl;
+using System.Security.Cryptography;
+using System.Security.Principal;
+
+namespace SharkCageInstallerHelper
+{
+    [RunInstaller(true)]
+    public partial class SharkCageInstallerHelper : Installer
+    {
+        const string service_name = "shark-cage-service";
+
+        public override void Install(System.Collections.IDictionary state_saver)
+        {
+            base.Install(state_saver);
+        }
+
+        public override void Commit(System.Collections.IDictionary saved_state)
+        {
+            base.Commit(saved_state);
+
+            string target_dir = Context.Parameters["DP_TargetDir"].ToString();
+
+            AssimilateConfigToCurrentSystem(target_dir);
+
+            InstallAndStartService(target_dir);
+        }
+
+        public override void Rollback(IDictionary saved_state)
+        {
+            base.Rollback(saved_state);
+            UninstallService();
+        }
+
+        public override void Uninstall(System.Collections.IDictionary saved_state)
+        {
+            base.Uninstall(saved_state);
+            UninstallService();
+        }
+
+        private void AssimilateConfigToCurrentSystem(string dir_path)
+        {
+            var config_path = Path.Combine(
+                Environment.GetFolderPath(Environment.SpecialFolder.CommonDocuments),
+                "SharkCage\\CageConfigurator.sconfig");
+
+            Registry.SetValue(@"HKEY_LOCAL_MACHINE\SOFTWARE\SharkCage\Configs", "CageConfigurator", config_path);
+
+            MoveConfigFileToConfigPath(config_path, dir_path);
+
+            UpdateContentOfConfigFile(config_path, dir_path);
+
+            UpdateSecurityAclsOfConfigFile(config_path);
+        }
+
+        private void MoveConfigFileToConfigPath(string config_path, string dir_path)
+        {
+            if (File.Exists(config_path))
+            {
+                File.Delete(config_path);
+            }
+
+            File.Move(Path.Combine(dir_path, "CageConfigurator.sconfig"), config_path);
+        }
+
+        private void UpdateContentOfConfigFile(string config_path, string dir_path)
+        {
+            const string APPLICATION_PATH_PROPERTY = "application_path";
+            const string APPLICATION_HASH_PROPERTY = "binary_hash";
+            var path_to_config_exe = Path.Combine(dir_path, "CageConfigurator.exe");
+
+            var json = JObject.Parse(File.ReadAllText(config_path));
+            var application_path = json.GetValue(APPLICATION_PATH_PROPERTY).ToString();
+            json[APPLICATION_PATH_PROPERTY] = path_to_config_exe;
+            json[APPLICATION_HASH_PROPERTY] = GetSha512Hash(path_to_config_exe);
```

Comment:
I'm not sure patching the hash on the target system is a good idea. We can generate the hash once before creating the installer and just patch the correct path

Reply:
The hash looks different on our machines with the same target (debug) so I thought it will be the same result with the release target. I don't know what but something on the system must have an influence on the calculation of the hash over the binary. So I think the only way is it to calculate the hash on the target system but I think that should not decrease the security at all especially because up to the point of installation, the config is writable.

Maybe we should create an issue for that to think about how we can install the cage in a secure way on a already infected system.

Reply:
Will create an issue for that.